crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1832,23 +1840,55 @@ impl<'db> ClassType<'db> {
     /// Note that any class (except for `object`) that has no explicit
     /// bases will implicitly inherit from `object` at runtime. Nonetheless,
     /// this method does *not* include `object` in the bases it iterates over.
-    fn explicit_bases(self, db: &'db dyn Db) -> impl Iterator<Item = Type<'db>> {
-        let definition = self.definition(db);
+    ///
+    /// ## Why is this a salsa query?
+    ///
+    /// This is a salsa query to short-circuit the invalidation
+    /// when the class's AST node changes.
+    ///
+    /// Were this not a salsa query, then the calling query
+    /// would depend on the class's AST and rerun for every change in that file.
+    fn explicit_bases(self, db: &'db dyn Db) -> &[Type<'db>] {
+        self.explicit_bases_query(db)
+    }
```

Comment:
FWIW I'm not sure I see the need for this wrapper method right now -- everything compiles for me with this diff applied to your branch. I also doubt that we'll start accessing the bases from more places in the future; in 99% of cases, you really want to iterate over the MRO rather than the bases. However, I don't feel strongly; I'm okay with keeping it for now.

```diff
diff --git a/crates/red_knot_python_semantic/src/types.rs b/crates/red_knot_python_semantic/src/types.rs
index b33d0fe4f..7ce85fb19 100644
--- a/crates/red_knot_python_semantic/src/types.rs
+++ b/crates/red_knot_python_semantic/src/types.rs
@@ -1835,7 +1835,7 @@ impl<'db> ClassType<'db> {
             })
     }
 
-    /// Return an iterator over the inferred types of this class's *explicit* bases.
+    /// Return the inferred types of this class's *explicit* bases.
     ///
     /// Note that any class (except for `object`) that has no explicit
     /// bases will implicitly inherit from `object` at runtime. Nonetheless,
@@ -1848,12 +1848,8 @@ impl<'db> ClassType<'db> {
     ///
     /// Were this not a salsa query, then the calling query
     /// would depend on the class's AST and rerun for every change in that file.
-    fn explicit_bases(self, db: &'db dyn Db) -> &[Type<'db>] {
-        self.explicit_bases_query(db)
-    }
-
     #[salsa::tracked(return_ref)]
-    fn explicit_bases_query(self, db: &'db dyn Db) -> Box<[Type<'db>]> {
+    fn explicit_bases(self, db: &'db dyn Db) -> Box<[Type<'db>]> {
         let class_stmt = self.node(db);
 
         if class_stmt.type_params.is_some() {
diff --git a/crates/red_knot_python_semantic/src/types/mro.rs b/crates/red_knot_python_semantic/src/types/mro.rs
index ea298478f..5af11eccd 100644
--- a/crates/red_knot_python_semantic/src/types/mro.rs
+++ b/crates/red_knot_python_semantic/src/types/mro.rs
@@ -40,7 +40,7 @@ impl<'db> Mro<'db> {
     }
 
     fn of_class_impl(db: &'db dyn Db, class: ClassType<'db>) -> Result<Self, MroErrorKind<'db>> {
-        let class_bases = class.explicit_bases(db);
+        let class_bases = &**class.explicit_bases(db);
 
         match class_bases {
             // `builtins.object` is the special case:
```

Reply:
I want to keep the `explicit_bases` function so that the return type is `&[Type]` over `&Box<[Type]>`

Reply:
I'm fine with that, but just so I understand fully: does that just give us better ergonomics? or is there a perf/generated code benefit?

Reply:
It's just ergonomics. A slice return type is more ergonomic than a `Box<[Type]>` (whether we use a `Box` or `Vec` internally is an implementation detail)